rasterio/_warp.pyx

```diff
@@ -317,42 +317,47 @@ def _reproject(
 
         log.debug("Created temp source dataset")
 
-        if src_transform:
-            for i in range(6):
-                gt[i] = src_transform[i]
-            retval = GDALSetGeoTransform(src_dataset, gt)
+        try:
+            src_osr = _osr_from_crs(src_crs)
+            OSRExportToWkt(src_osr, &srcwkt)
 
-            log.debug("Set transform on temp source dataset: %d", retval)
+            if src_transform:
+                for i in range(6):
+                    gt[i] = src_transform[i]
 
-            try:
-                src_osr = _osr_from_crs(src_crs)
-                OSRExportToWkt(src_osr, &srcwkt)
-                GDALSetProjection(src_dataset, srcwkt)
+                if not GDALError.none == GDALSetGeoTransform(src_dataset, gt):
+                    raise ValueError(
+                        "Failed to set transform on temp source dataset.")
+
+                if not GDALError.none == GDALSetProjection(
+                        src_dataset, srcwkt):
+                    raise ("Failed to set projection on temp source dataset.")
```

Comment:
We've got a new style for working with the GDAL error system. Wrap the calls to GDALSetGeoTransform and GDALSetProjection with a call to exc_wrap_int like `exc_wrap_int(GDALSetGeoTransform(...))` and the error codes will be checked and exceptions raised for you.